Add shp2csv example to experiment with the shapelib API

Notgnoshi opened the pull request:
```
$ ./build/examples/shp2csv/shp2csv -i data/mexico/roads | head -n 2
id,type,parts,vertices,min_x,max_x,min_y,max_y,min_z,max_z,min_m,max_m,geometry
0,3,1,4,-99.00114440917969,-98.13920593261719,22.55204772949219,23.31682777404785,0,0,0,0,"LINESTRING (-98.13920593261719 22.55204772949219, -98.80698394775391 23.24045944213867, -98.86305999755859 23.28426361083984, -99.00114440917969 23.31682777404785)"
```

Notgnoshi committed:
shp2csv: Add shapelib example for API comparison

shp2csv will be an example commandline tool to read a shapefile with
shapelib (without shpxx) and dump its geometries and attributes in a CSV
format like

    id,geometry,weight,name
    0,"POINT(0 0)",0.0,"P1"
    1,"POINT(1 1)",0.0,"P2"
    ...

Geometries and string attributes will always quoted. String attributes
will be parsed and any contained quotes will be escaped. The "id" and
"geometry" columns will always be present.

The purpose of the shp2csv example is to get a feel for what using "raw"
shapelib is like, and to provide a baseline to compare the TBD shpxx API
against.

Notgnoshi committed:
shp2csv: Read features and their IDs

Ipiano commented on examples/shp2csv/shp2csv.cpp:
```diff
@@ -0,0 +1,671 @@
+    g_verbose = options.verbose;
+    options.output << std::setprecision(16);
+
+    const SHPHandle shapefile = SHPOpen(options.input_filename.c_str(), "rb");
```
Should check for a nullptr here

Notgnoshi commented:
```sh
$ ./build/examples/shp2csv/shp2csv -i data/shape_eg_data/3dpoints
```

Gives

|id |type|parts|vertices|min_x              |max_x              |min_y            |max_y            |min_z|max_z|min_m|max_m|geometry                                         |
|---|----|-----|--------|-------------------|-------------------|-----------------|-----------------|-----|-----|-----|-----|-------------------------------------------------|
|0  |11  |0    |1       |0.4063999999999997 |0.4063999999999997 |7.484799999999999|7.484799999999999|0    |0    |0    |0    |POINT Z (0.4063999999999997 7.484799999999999 0) |
|1  |11  |0    |1       |2.5184             |2.5184             |7.484799999999999|7.484799999999999|0    |0    |0    |0    |POINT Z (2.5184 7.484799999999999 0)             |
|2  |11  |0    |1       |6.3904             |6.3904             |7.414400000000001|7.414400000000001|0    |0    |0    |0    |POINT Z (6.3904 7.414400000000001 0)             |
|3  |11  |0    |1       |9.558399999999999  |9.558399999999999  |7.343999999999999|7.343999999999999|0    |0    |0    |0    |POINT Z (9.558399999999999 7.343999999999999 0)  |
|4  |11  |0    |1       |0.05439999999999967|0.05439999999999967|4.175999999999999|4.175999999999999|0    |0    |0    |0    |POINT Z (0.05439999999999967 4.175999999999999 0)|
|5  |0   |0    |0       |0                  |0                  |0                |0                |0    |0    |0    |0    |NULL                                             |
|6  |0   |0    |0       |0                  |0                  |0                |0                |0    |0    |0    |0    |NULL                                             |
|7  |11  |0    |1       |9.769600000000001  |9.769600000000001  |3.612799999999999|3.612799999999999|0    |0    |0    |0    |POINT Z (9.769600000000001 3.612799999999999 0)  |

Notice points 5 and 6 are NULL.

But Poke from https://github.com/Notgnoshi/shpxx/pull/13 gives

```
ShpFile {
  file_header=ShpFileHeader {
    file_code=0x0000270aU,
    reserved1=0x00000000U,
    reserved2=0x00000000U,
    reserved3=0x00000000U,
    reserved4=0x00000000U,
    reserved5=0x00000000U,
    file_length=0x000001beU,
    version=0x000003e8U,
    shape_type=0x0000000bU,
    bounding_box=struct {
      x_min=#<-3.163473079561041>,
      y_min=#<-0.603677477660002>,
      x_max=#<11.925690157731081>,
      y_max=#<10.542566872427980>,
      z_min=#<0.000000000000000>,
      z_max=#<0.000000000000000>,
      m_min=#<0.000000000000000>,
      m_max=#<0.000000000000000>
    }
  },
  points=[ShpPointRecord {
    header=ShpRecordHeader {
      id=0x00000001U,
      length=0x0000000eU
    },
    shape_type=0x0000000bU,
    content=struct {
      point=ShpPoint {
        x=#<0.406400000000000>,
        y=#<7.484799999999999>
      },
      z=#<0.000000000000000>
    }
  },ShpPointRecord {
    header=ShpRecordHeader {
      id=0x00000002U,
      length=0x0000000eU
    },
    shape_type=0x0000000bU,
    content=struct {
      point=ShpPoint {
        x=#<2.518400000000000>,
        y=#<7.484799999999999>
      },
      z=#<0.000000000000000>
    }
  },ShpPointRecord {
    header=ShpRecordHeader {
      id=0x00000003U,
      length=0x0000000eU
    },
    shape_type=0x0000000bU,
    content=struct {
      point=ShpPoint {
        x=#<6.390400000000000>,
        y=#<7.414400000000001>
      },
      z=#<0.000000000000000>
    }
  },ShpPointRecord {
    header=ShpRecordHeader {
      id=0x00000004U,
      length=0x0000000eU
    },
    shape_type=0x0000000bU,
    content=struct {
      point=ShpPoint {
        x=#<9.558399999999999>,
        y=#<7.343999999999999>
      },
      z=#<0.000000000000000>
    }
  },ShpPointRecord {
    header=ShpRecordHeader {
      id=0x00000005U,
      length=0x0000000eU
    },
    shape_type=0x0000000bU,
    content=struct {
      point=ShpPoint {
        x=#<0.054400000000000>,
        y=#<4.175999999999999>
      },
      z=#<0.000000000000000>
    }
  },ShpPointRecord {
    header=ShpRecordHeader {
      id=0x00000006U,
      length=0x0000000eU
    },
    shape_type=0x0000000bU,
    content=struct {
      point=ShpPoint {
        x=#<2.870400000000000>,
        y=#<3.894400000000000>
      },
      z=#<0.000000000000000>
    }
  },ShpPointRecord {
    header=ShpRecordHeader {
      id=0x00000007U,
      length=0x0000000eU
    },
    shape_type=0x0000000bU,
    content=struct {
      point=ShpPoint {
        x=#<6.953600000000002>,
        y=#<4.316800000000000>
      },
      z=#<0.000000000000000>
    }
  },ShpPointRecord {
    header=ShpRecordHeader {
      id=0x00000008U,
      length=0x0000000eU
    },
    shape_type=0x0000000bU,
    content=struct {
      point=ShpPoint {
        x=#<9.769600000000001>,
        y=#<3.612799999999999>
      },
      z=#<0.000000000000000>
    }
  },ShpPointRecord {
    header=ShpRecordHeader {
      id=0x00000009U,
      length=0x0000000eU
    },
    shape_type=0x0000000bU,
    content=struct {
      point=ShpPoint {
        x=#<1.110400000000000>,
        y=#<1.289600000000000>
      },
      z=#<0.000000000000000>
    }
  },ShpPointRecord {
    header=ShpRecordHeader {
      id=0x0000000aU,
      length=0x0000000eU
    },
    shape_type=0x0000000bU,
    content=struct {
      point=ShpPoint {
        x=#<5.616000000000000>,
        y=#<1.078400000000000>
      },
      z=#<0.000000000000000>
    }
  },ShpPointRecord {
    header=ShpRecordHeader {
      id=0x0000000bU,
      length=0x0000000eU
    },
    shape_type=0x0000000bU,
    content=struct {
      point=ShpPoint {
        x=#<9.840000000000000>,
        y=#<0.867200000000000>
      },
      z=#<0.000000000000000>
    }
  },ShpPointRecord {
    header=ShpRecordHeader {
      id=0x0000000cU,
      length=0x0000000eU
    },
    shape_type=0x0000000bU,
    content=struct {
      point=ShpPoint {
        x=#<-1.049999999999998>,
        y=#<8.603703703703703>
      },
      z=#<2.000000000000000>
    }
  },ShpPointRecord {
    header=ShpRecordHeader {
      id=0x0000000dU,
      length=0x0000000eU
    },
    shape_type=0x0000000bU,
    content=struct {
      point=ShpPoint {
        x=#<2.209259259259261>,
        y=#<8.603703703703703>
      },
      z=#<2.000000000000000>
    }
  },ShpPointRecord {
    header=ShpRecordHeader {
      id=0x0000000eU,
      length=0x0000000eU
    },
    shape_type=0x0000000bU,
    content=struct {
      point=ShpPoint {
        x=#<7.994444444444445>,
        y=#<8.685185185185183>
      },
      z=#<2.000000000000000>
    }
  },ShpPointRecord {
    header=ShpRecordHeader {
      id=0x0000000fU,
      length=0x0000000eU
    },
    shape_type=0x0000000bU,
    content=struct {
      point=ShpPoint {
        x=#<10.601851851851853>,
        y=#<8.603703703703703>
      },
      z=#<2.000000000000000>
    }
  },ShpPointRecord {
    header=ShpRecordHeader {
      id=0x00000010U,
      length=0x0000000eU
    },
    shape_type=0x0000000bU,
    content=struct {
      point=ShpPoint {
        x=#<-1.049999999999998>,
        y=#<-0.603703703703703>
      },
      z=#<0.000000000000000>
    }
  },ShpPointRecord {
    header=ShpRecordHeader {
      id=0x00000011U,
      length=0x0000000eU
    },
    shape_type=0x0000000bU,
    content=struct {
      point=ShpPoint {
        x=#<5.142592592592594>,
        y=#<-0.603703703703703>
      },
      z=#<0.000000000000000>
    }
  },ShpPointRecord {
    header=ShpRecordHeader {
      id=0x00000012U,
      length=0x0000000eU
    },
    shape_type=0x0000000bU,
    content=struct {
      point=ShpPoint {
        x=#<10.683333333333334>,
        y=#<-0.359259259259259>
      },
      z=#<0.000000000000000>
    }
  },ShpPointRecord {
    header=ShpRecordHeader {
      id=0x00000013U,
      length=0x0000000eU
    },
    shape_type=0x0000000bU,
    content=struct {
      point=ShpPoint {
        x=#<-3.163473079561041>,
        y=#<10.542566872427980>
      },
      z=#<3.000000000000000>
    }
  },ShpPointRecord {
    header=ShpRecordHeader {
      id=0x00000014U,
      length=0x0000000eU
    },
    shape_type=0x0000000bU,
    content=struct {
      point=ShpPoint {
        x=#<2.377079046639233>,
        y=#<10.306798696844989>
      },
      z=#<3.000000000000000>
    }
  },ShpPointRecord {
    header=ShpRecordHeader {
      id=0x00000015U,
      length=0x0000000eU
    },
    shape_type=0x0000000bU,
    content=struct {
      point=ShpPoint {
        x=#<9.096472050754457>,
        y=#<10.542566872427980>
      },
      z=#<3.000000000000000>
    }
  },ShpPointRecord {
    header=ShpRecordHeader {
      id=0x00000016U,
      length=0x0000000eU
    },
    shape_type=0x0000000bU,
    content=struct {
      point=ShpPoint {
        x=#<11.925690157750340>,
        y=#<10.188914609053494>
      },
      z=#<3.000000000000000>
    }
  }]
}
```

Notice there are 22 points, and that points 5 and 6 are _not_ NULL.

Notgnoshi committed:
Add support for various sanitizers

Notgnoshi commented on CMakeLists.txt:
```diff
@@ -68,7 +73,25 @@ endif()
+if(SHPXX_INSTRUMENT_UBSAN)
+    add_compile_options(-g3 -fsanitize=undefined)
+    add_link_options(-fsanitize=undefined)
+endif()
```
The other option for this that I like is to add https://github.com/arsenm/sanitizers-cmake as a submodule

Notgnoshi committed:
Update Findshapelib for shapelib 1.6.3

Notgnoshi committed:
shp2csv: Fix MULTIPOINT parsing

multipoints don't use parts, just nVertices.

Notgnoshi committed:
shp2csv: Fix polygon WKT formatting

Single-ring polygons are still POLYGON((...)) not POLYGON(...)

Notgnoshi committed:
shp2csv: Use winding order to determine polygon ring type

Multiple exterior rings => MULTIPOLYGON
Single exterior ring, but multiple holes => POLYGON with holes

Notgnoshi committed:
shp2csv: Parse DBF schema and include empty fields in CSV

Notgnoshi committed:
shp2csv: DBF value parsing

Notgnoshi commented:
> ```shell
> $ ./build/examples/shp2csv/shp2csv -i data/shape_eg_data/3dpoints
> ```
> 
> Gives
> id 	type 	parts 	vertices 	min_x 	max_x 	min_y 	max_y 	min_z 	max_z 	min_m 	max_m 	geometry
> 0 	11 	0 	1 	0.4063999999999997 	0.4063999999999997 	7.484799999999999 	7.484799999999999 	0 	0 	0 	0 	POINT Z (0.4063999999999997 7.484799999999999 0)
> 1 	11 	0 	1 	2.5184 	2.5184 	7.484799999999999 	7.484799999999999 	0 	0 	0 	0 	POINT Z (2.5184 7.484799999999999 0)
> 2 	11 	0 	1 	6.3904 	6.3904 	7.414400000000001 	7.414400000000001 	0 	0 	0 	0 	POINT Z (6.3904 7.414400000000001 0)
> 3 	11 	0 	1 	9.558399999999999 	9.558399999999999 	7.343999999999999 	7.343999999999999 	0 	0 	0 	0 	POINT Z (9.558399999999999 7.343999999999999 0)
> 4 	11 	0 	1 	0.05439999999999967 	0.05439999999999967 	4.175999999999999 	4.175999999999999 	0 	0 	0 	0 	POINT Z (0.05439999999999967 4.175999999999999 0)
> 5 	0 	0 	0 	0 	0 	0 	0 	0 	0 	0 	0 	NULL
> 6 	0 	0 	0 	0 	0 	0 	0 	0 	0 	0 	0 	NULL
> 7 	11 	0 	1 	9.769600000000001 	9.769600000000001 	3.612799999999999 	3.612799999999999 	0 	0 	0 	0 	POINT Z (9.769600000000001 3.612799999999999 0)
> 
> Notice points 5 and 6 are NULL.
> 

Looked into this a bit more. The `.shp` and `.dbf` both look to have 22 non-null fields, but the `.shx` truncates the number of records, and has incorrect offsets for fields 5 and 6.

So shp2csv is parsing it correctly, but the .shx index is corrupted. We *could* detect this and rebuild the .shx index with SHPRestoreSHX, but I'm going to claim that's out-of-scope.

Notgnoshi committed:
Fix clang-tidy version

The lint.yml workflow installs clang-tidy-14
